services/pull/merge_squash.go

```diff
@@ -7,24 +7,58 @@ import (
 	"fmt"
 
 	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
 	"code.gitea.io/gitea/modules/git"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
 )
 
+// doMergeStyleSquash gets a commit author signature for squash commits
+func getAuthorSignatureSquash(ctx *mergeContext) (*git.Signature, error) {
+	if err := ctx.pr.Issue.LoadPoster(ctx); err != nil {
+		return nil, fmt.Errorf("LoadPoster: %w", err)
+	}
+
+	// Try to get an signature from the same user in one of the commits, as the
+	// poster email might be private or commits might have a different signature
+	// than the primary email address of the poster.
+	gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, ctx.tmpBasePath)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to open repository: Error %v", err)
+	}
+	defer closer.Close()
+
+	commits, err := gitRepo.CommitsBetweenIDs(trackingBranch, "HEAD")
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get commits between: %s %s Error %v", "HEAD", trackingBranch, err)
+	}
+
+	for _, commit := range commits {
+		if commit.Author != nil {
+			commitUser, _ := user_model.GetUserByEmail(ctx, commit.Author.Email)
```

Comment:
We could cache the users by email so that we could not load it from database again.